### PR TITLE
Convert more of pkg/systemd to functional components

### PR DESCRIPTION
**Pull request by jelly**
Obviously it's a bit wasteful to convert everything without bringing in improvements but most of this was quite easy to convert.

### CODE REVIEW

File: pkg/systemd/hwinfo.jsx [hunk 1]
```diff
@@ -55,78 +55,74 @@ import "./hwinfo.scss";
...
+                            <DescriptionListTerm>{ _("BIOS") }</DescriptionListTerm>
+                            <DescriptionListDescription>{ info.bios_vendor }</DescriptionListDescription>
+                        </DescriptionListGroup>
+                        <DescriptionListGroup>
+                            <DescriptionListTerm>{ _("BIOS version") }</DescriptionListTerm>
+                            <DescriptionListDescription>{ info.bios_version }</DescriptionListDescription>
+                        </DescriptionListGroup>
+                        <DescriptionListGroup>
+                            <DescriptionListTerm>{ _("BIOS date") }</DescriptionListTerm>
+                            <DescriptionListDescription>{ bios_date ? timeformat.date(bios_date) : info.bios_date }</DescriptionListDescription>
```
Comment (cockpituous) [inline pkg/systemd/hwinfo.jsx:108]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19378-20230922-161344-dd3ec708-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/hwinfo.jsx [hunk 2]
```diff
@@ -192,173 +183,164 @@ class CPUSecurityMitigationsDialog extends React.Component {
         cockpit.script(kernelopt_sh, options, { superuser: "require", err: "message" })
                 .then(() => {
                     cockpit.spawn(["shutdown", "--reboot", "now"], { superuser: "require", err: "message" })
-                            .catch(error => this.setState({ rebooting: false, alert: error.message }));
+                            .catch(error => { setRebooting(true); setAlert(error.message) });
```
Comment (cockpituous) [inline pkg/systemd/hwinfo.jsx:186]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19378-20230922-161344-dd3ec708-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/hwinfo.jsx [hunk 3]
```diff
@@ -192,173 +183,164 @@ class CPUSecurityMitigationsDialog extends React.Component {
...
-                </Button>
+                <Text className='cpu-mitigations-dialog-info' component={TextVariants.p}>
+                    { _("Software-based workarounds help prevent CPU security issues. These mitigations have the side effect of reducing performance. Change these settings at your own risk.") }
+                </Text>
+                <DataList>
+                    { rows }
+                </DataList>
+                { alert !== undefined &&
+                <Alert variant="danger"
+                    actionClose={<AlertActionCloseButton onClose={() => setAlert(undefined)} />}
```
Comment (cockpituous) [inline pkg/systemd/hwinfo.jsx:250]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19378-20230922-161344-dd3ec708-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/hwinfo.jsx [hunk 4]
```diff
@@ -192,173 +183,164 @@ class CPUSecurityMitigationsDialog extends React.Component {
...
+                columns={ [_("ID"), _("Memory technology"), _("Type"), _("Size"), _("State"), _("Rank"), _("Speed")]}
+                rows={ info.memory.map(dimm => ({
+                    props: { key: dimm.locator },
+                    columns: [dimm.locator, dimm.technology, dimm.type, dimm.size, dimm.state, dimm.rank, dimm.speed]
+                })) } />
         );
+    } else if (!superuser.allowed) {
+        memory = (<EmptyState>
+            {_("Viewing memory information requires administrative access.")}
+        </EmptyState>);
```
Comment (cockpituous) [inline pkg/systemd/hwinfo.jsx:299]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19378-20230922-161344-dd3ec708-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/overview-cards/healthCard.jsx [hunk 5]
```diff
@@ -31,22 +31,20 @@ import "./healthCard.scss";
...
-                    </ul>
-                </CardBody>
-                <CardFooter />
-            </Card>
-        );
-    }
-}
+export const HealthCard = () => {
+    return (
+        <Card className="system-health">
```
Comment (martinpitt) [inline pkg/systemd/overview-cards/healthCard.jsx:36]: Style issue: For such a "simple mapping" component, return the element directly: `HealthCard = () => <Card>`, or if you want to keep the (admittedly nicer) indendation on the next line, use `HealthCard = () => (` and save both the extra `{ return }` and the extra indendation.

File: pkg/systemd/hwinfo.jsx [hunk 6]
```diff
@@ -154,25 +154,20 @@ function availableMitigations() {
...
 
-    constructor(props) {
-        super(props);
-        this.saveAndReboot = this.saveAndReboot.bind(this);
-        this.state = {
-            nosmt: undefined,
-            alert: undefined,
-            rebooting: false,
-        };
+    React.useEffect(() => {
```
Comment (martinpitt) [inline pkg/systemd/hwinfo.jsx:162]: I'd rather we use `useInit` for that -- it's more abstract and better communicates what it does. useEffect() can be used for lots of other things.

**Review by martinpitt — APPROVED** [review]
Thanks! If you want to get it out of your eyes quickly, good enough. I have two improvement requests, though.

File: pkg/systemd/hwinfo.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply pkg/systemd/hwinfo.jsx:186]: Very well spotted, thanks!

File: pkg/systemd/overview-cards/healthCard.jsx (hunk 5 above)
Comment (jelly) [reply pkg/systemd/overview-cards/healthCard.jsx:36]: Excellent point!

File: pkg/systemd/hwinfo.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (jelly) [reply pkg/systemd/hwinfo.jsx:186]: Ah no, it is not! Nicely spotted!

File: pkg/systemd/hwinfo.jsx [hunk 7]
```diff
@@ -43,90 +43,87 @@ import { Switch } from "@patternfly/react-core/dist/esm/components/Switch/index.
...
+                            <DescriptionListTerm>{ _("BIOS") }</DescriptionListTerm>
+                            <DescriptionListDescription>{ info.bios_vendor }</DescriptionListDescription>
+                        </DescriptionListGroup>
+                        <DescriptionListGroup>
+                            <DescriptionListTerm>{ _("BIOS version") }</DescriptionListTerm>
+                            <DescriptionListDescription>{ info.bios_version }</DescriptionListDescription>
+                        </DescriptionListGroup>
+                        <DescriptionListGroup>
+                            <DescriptionListTerm>{ _("BIOS date") }</DescriptionListTerm>
+                            <DescriptionListDescription>{ bios_date ? timeformat.date(bios_date) : info.bios_date }</DescriptionListDescription>
```
Comment (cockpituous) [inline pkg/systemd/hwinfo.jsx:109]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19378-20230925-080729-1e17ab19-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/hwinfo.jsx [hunk 8]
```diff
@@ -192,173 +184,164 @@ class CPUSecurityMitigationsDialog extends React.Component {
         cockpit.script(kernelopt_sh, options, { superuser: "require", err: "message" })
                 .then(() => {
                     cockpit.spawn(["shutdown", "--reboot", "now"], { superuser: "require", err: "message" })
-                            .catch(error => this.setState({ rebooting: false, alert: error.message }));
+                            .catch(error => { setRebooting(false); setAlert(error.message) });
```
Comment (cockpituous) [inline pkg/systemd/hwinfo.jsx:187]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19378-20230925-080729-1e17ab19-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/hwinfo.jsx [hunk 9]
```diff
@@ -192,173 +184,164 @@ class CPUSecurityMitigationsDialog extends React.Component {
...
-                </Button>
+                <Text className='cpu-mitigations-dialog-info' component={TextVariants.p}>
+                    { _("Software-based workarounds help prevent CPU security issues. These mitigations have the side effect of reducing performance. Change these settings at your own risk.") }
+                </Text>
+                <DataList>
+                    { rows }
+                </DataList>
+                { alert !== undefined &&
+                <Alert variant="danger"
+                    actionClose={<AlertActionCloseButton onClose={() => setAlert(undefined)} />}
```
Comment (cockpituous) [inline pkg/systemd/hwinfo.jsx:251]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19378-20230925-080729-1e17ab19-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/hwinfo.jsx [hunk 10]
```diff
@@ -192,173 +184,164 @@ class CPUSecurityMitigationsDialog extends React.Component {
...
+                columns={ [_("ID"), _("Memory technology"), _("Type"), _("Size"), _("State"), _("Rank"), _("Speed")]}
+                rows={ info.memory.map(dimm => ({
+                    props: { key: dimm.locator },
+                    columns: [dimm.locator, dimm.technology, dimm.type, dimm.size, dimm.state, dimm.rank, dimm.speed]
+                })) } />
         );
+    } else if (!superuser.allowed) {
+        memory = (<EmptyState>
+            {_("Viewing memory information requires administrative access.")}
+        </EmptyState>);
```
Comment (cockpituous) [inline pkg/systemd/hwinfo.jsx:300]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19378-20230925-080729-1e17ab19-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/hwinfo.jsx [hunk 11]
```diff
@@ -154,25 +155,20 @@ function availableMitigations() {
...
-            alert: undefined,
-            rebooting: false,
-        };
+    useInit(() => {
         availableMitigations().then(({ available, nosmt_enabled }) => {
-            this.setState({ nosmt: nosmt_enabled });
+            setNoSMT(nosmt_enabled);
         });
-    }
+    }, []);
```
Comment (martinpitt) [inline pkg/systemd/hwinfo.jsx:167]: We don't usually give dependencies for `useInit()`, but good enough.

**Review by martinpitt — DISMISSED** [review]
Cheers!

**Review by martinpitt — APPROVED** [review]
Heh, thanks. Would really have been good enough, though :grin:

**Pull request merged**